hal/api/EventAdapter.h

```diff
@@ -0,0 +1,302 @@
+        construct_callback(callback(func, arg));
+    }
+
+    /** Create an EventAdapter with a static function and bound pointer callback
+     *  @param func     Static function to attach
+     *  @param arg      Pointer argument to function
+     */
+    template<typename T>
+    EventAdapter(void (*func)(const volatile T*), const volatile T *arg) {
+        construct_callback(callback(func, arg));
+    }
+
+    /** Create an EventAdapter that holds an Event
+     *  @param evt The event
+     */
+    EventAdapter(const adapter_event_t& evt) {
+        construct_event(evt);
+    }
+
+    EventAdapter(const EventAdapter& adapter) {
+        _attach(adapter);
+    }
+
+    /** Test if a function is attached to this adapter
+     */
+    operator bool() const {
+        if (_is_cb) {
+            return get_callback()->operator bool();
+        } else {
+            return true;
+        }
+    }
+
+    EventAdapter& operator =(const EventAdapter& rhs) {
+        if (this != &rhs) {
+            destroy();
+            _attach(rhs);
+        }
+        return *this;
+    }
+
+    /** Attach this adapter to the callback or event in a different adapter
+     *  @param adapter The EventAdapter to attach.
+     */
+    void attach(const EventAdapter& adapter) {
+        destroy();
+        _attach(adapter);
+    }
+
+    /** Call the callback or event in this adapter
+     */
+    void call() {
+        if (_is_cb) {
+            get_callback()->call();
+        } else {
+            get_event()->call();
+        }
+    }
+
+    /** Call the callback or event in this adapter
+     */
+    void operator ()() {
+        call();
+    }
+
+    ~EventAdapter() {
+        destroy();
+    }
+
+private:
+    // No unions for non-POD data types in C++03. Need to use placement new to simulate the
+    // union between a Callback<void()> and an Event<void()>
+    union {
+        // Ensure alignment by declaring the storage uint32_t instead of uint8_t
+        uint32_t _cb_data[(sizeof(adapter_callback_t) + sizeof(uint32_t) - 1) / sizeof(uint32_t)];
+        uint32_t _evt_data[(sizeof(adapter_event_t) + sizeof(uint32_t) - 1) / sizeof(uint32_t)];
+    };
+    bool  _is_cb;
+
+    void destroy() {
+        if (_is_cb) {
+            get_callback()->~adapter_callback_t();
+        } else {
+            get_event()->~adapter_event_t();
+        }
+    }
+
+    void construct_callback(const adapter_callback_t& cb) {
+        _is_cb = true;
+        adapter_callback_t *p_cb = reinterpret_cast<adapter_callback_t*>(&_cb_data);
+        new (p_cb) adapter_callback_t();
+        p_cb->attach(cb);
+    }
+
+    void construct_event(const adapter_event_t&evt) {
+        _is_cb = false;
+        adapter_event_t *p_evt = reinterpret_cast<adapter_event_t*>(&_evt_data);
+        new(p_evt) adapter_event_t(evt);
+    }
+
+    const adapter_event_t *get_event() const {
+        return reinterpret_cast<const adapter_event_t*>(&_evt_data);
```

Comment:
It's a small thing, but these functions look like they really want an assert on `_is_cb` : )